File: Koko Eating Bananas - Leetcode 875.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def k_works(k):
            hours = 0

            for p in piles:
                hours += ceil(p / k)

            return hours <= h

        l = 1
        r = max(piles) # m

        while l < r:
            k = (l + r) // 2

            if k_works(k):
                r = k
            else:
                l = k + 1

        return r

# Time Complexity: O(n * log(m))
# Space Complexity: O(1)

File: Koko Eating Bananas - Leetcode 875/Koko Eating Bananas - Leetcode 875.cpp
#include <vector>
#include <algorithm>
#include <cmath>

class Solution {
public:
    int minEatingSpeed(std::vector<int>& piles, int h) {
        int left = 1;
        int right = *std::max_element(piles.begin(), piles.end());

        while (left < right) {
            int mid = left + (right - left) / 2;

            if (canFinish(piles, h, mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }

private:
    bool canFinish(const std::vector<int>& piles, int h, int k) {
        int hours = 0;
        for (int pile : piles) {
            hours += std::ceil(static_cast<double>(pile) / k);
        }
        return hours <= h;
    }
};
